keps/0015-dry-run.md

```diff
@@ -53,10 +53,52 @@ either as an argument for built-in admission plugins or through a dryRun
 query-parameter for dynamic webhooks, will give them a chance not to have any
 side-effect.
 
-All admission controllers will have to be verified and changed. A new flag will
-be added so that webhooks can explicitly register with dry-run support. If
-dry-run is requested on a non-supported webhook, the request will be completely
-rejected.
+All built-in admission controllers will have to be checked, and the ones with side
+effects will have to be skipped or changed to only have side effects in the non
+dry-run case. Some examples of built-in admission controllers with the possibility
+forside effects are:
+- ResourceQuota
+- EventRateLimit
+- NamespaceAutoProvision
+- (Valid|Mut)atingAdmissionWebhook
+
+To address the possibility of webhook authors relying on [side effects](https://github.com/kubernetes/website/blame/836629cb118e0f74545cc7d6d97aa6b9edfa1a16/content/en/docs/reference/access-authn-authz/admission-controllers.md#L582-L584), a new field
+will be added to ```admissionregistration.k8s.io/v1beta1.ValidatingWebhookConfiguration``` and
+```admissionregistration.k8s.io/v1beta1.MutatingWebhookConfiguration``` so that webhooks
+can explicitly register as having dry-run support.
+If dry-run is requested on a non-supported webhook, the request will be completely rejected,
+as a 400: Bad Request, and the name of the webhook/webhooks which don't support dry run will be
+listed. This field will be defaulted to true and deprecated in v1, and completely removed in v2.
+All webhooks registered with v2 will be assumed to support dry run.
+```golang
+// admissionregistration.k8s.io/v1beta1
+...
+type Webhook struct {
+	...
+	// SupportsDryRun defines whether this webhook will correctly handle dryRun requests.
+	// If false, any dryRun requests to resources/subresources this webhook applies to
+	// will be completely rejected and the webhook will not be called.
+	// Defaults to false.
+	// +optional
+	SupportsDryRun *bool `json:"supportsDryRun,omitempty" protobuf:"varint,6,number,opt,name=supportsDryRun"`
+}
+```
+
+Additionally, a new field will be added to ```admission.k8s.io/v1beta1.AdmissionReview```
+API object, indicating whether or not the request being reviewed is for a dry-run:
+```golang
+// admission.k8s.io/v1beta1
+...
+type AdmissionRequest struct {
+	...
+	// DryRun indicates that modifications will definitely not be persisted for this request.
+	// Defaults to false.
+	// +optional
+	DryRun *bool `json:"dryRun,omitempty" protobuf:"varint,11,number,opt,name=dryRun"`
+}
+```
+
+
```

Comment:
also need an update to the `admission.Attributes` interface (e.g. `DryRun() bool`)

Reply:
I wasn't sure if that goes in the KEP because it wasn't an API change, but since it's an interface people could import I see how it could still go here.